### ISSUE TITLE
Illegal aliasing is allowed by AntiAliasing

### ISSUE THREAD

**Issue by brunnerant** [epfl-lara/stainless#917]
`AntiAliasing` is supposed to report illegal uses of aliasing and crash Stainless in the later case. If I am not mistaken, it is pretty permissive and supports functions returning aliased results if they are non-recursive (it handles those by inlining the function to detect the exact aliasing).
But on the following example, it doesn't report any illegal aliasing, when it actually should:
```Scala
object Bug {
  case class A(var a: BigInt)

  def rec(a1: A, a2: A, n: BigInt): A = {
    if (n == 0) a1
    else rec(a2, a1, n - 1)
  }

  def bug() = {
    val a1 = A(1)
    val a2 = A(2)
    val a = rec(a1, a2, 5)
    a.a = 3
  }
}
```
The code gets translated to (after `AntiAliasing`):
```Scala
def bug: Unit = {
  var a1: A = A(1)
  var a2: A = A(2)
  var a: A = rec(a1, a2, 5)
  a = A(3)
  ()
}
```
Which is wrong, since `a1` and `a2` should be (conditonally) mutated.

**Comment by jad-hamza** [comment]
Thanks for the report! Can you post the full code with `A`?

**Comment by brunnerant** [comment]
I think that the issue might be coming from [there](https://github.com/epfl-lara/stainless/blob/master/core/src/main/scala/stainless/extraction/imperative/EffectsAnalyzer.scala#L353). Instead of returning an empty set of effects, I think that we should be throwing an exception indicating that the effects of a recursive functions are not known. I tried changing it and it does not seem to break the integration tests, and it makes the specific example that I posted crash, which is what we want.
